File: animated-drawable/src/main/java/com/facebook/fresco/animation/bitmap/cache/KeepLastFrameCache.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
package com.facebook.fresco.animation.bitmap.cache;

import android.graphics.Bitmap;
import com.facebook.common.references.CloseableReference;
import com.facebook.fresco.animation.bitmap.BitmapAnimationBackend;
import com.facebook.fresco.animation.bitmap.BitmapFrameCache;
import com.facebook.imageutils.BitmapUtil;
import javax.annotation.Nullable;
import javax.annotation.concurrent.GuardedBy;

/**
 * Simple bitmap cache that keeps the last frame and reuses it if possible.
 */
public class KeepLastFrameCache implements BitmapFrameCache {

  private static final int FRAME_NUMBER_UNSET = -1;

  private int mLastFrameNumber = FRAME_NUMBER_UNSET;

  @Nullable
  private FrameCacheListener mFrameCacheListener;

  @GuardedBy("this")
  @Nullable
  private CloseableReference<Bitmap> mLastBitmapReference;

  @Nullable
  @Override
  public synchronized CloseableReference<Bitmap> getCachedFrame(int frameNumber) {
    if (mLastFrameNumber == frameNumber) {
      return CloseableReference.cloneOrNull(mLastBitmapReference);
    }
    return null;
  }

  @Nullable
  @Override
  public synchronized CloseableReference<Bitmap> getFallbackFrame(int frameNumber) {
    return CloseableReference.cloneOrNull(mLastBitmapReference);
  }

  @Override
  public synchronized CloseableReference<Bitmap> getBitmapToReuseForFrame(
      int frameNumber,
      int width,
      int height) {
    try {
      return CloseableReference.cloneOrNull(mLastBitmapReference);
    } finally {
      closeAndResetLastBitmapReference();
    }
  }

  @Override
  public synchronized boolean contains(int frameNumber) {
    return frameNumber == mLastFrameNumber && CloseableReference.isValid(mLastBitmapReference);
  }

  @Override
  public synchronized int getSizeInBytes() {
    return mLastBitmapReference == null
        ? 0
        : BitmapUtil.getSizeInBytes(mLastBitmapReference.get());
  }

  @Override
  public synchronized void clear() {
    closeAndResetLastBitmapReference();
  }

  @Override
  public synchronized void onFrameRendered(
      int frameNumber,
      CloseableReference<Bitmap> bitmapReference,
      @BitmapAnimationBackend.FrameType int frameType) {
    if (bitmapReference != null
        && mLastBitmapReference != null
        && bitmapReference.get().equals(mLastBitmapReference.get())) {
      return;
    }
    CloseableReference.closeSafely(mLastBitmapReference);
    if (mFrameCacheListener != null && mLastFrameNumber != FRAME_NUMBER_UNSET) {
      mFrameCacheListener.onFrameEvicted(this, mLastFrameNumber);
    }
    mLastBitmapReference = CloseableReference.cloneOrNull(bitmapReference);
    if (mFrameCacheListener != null) {
      mFrameCacheListener.onFrameCached(this, frameNumber);
    }
    mLastFrameNumber = frameNumber;
  }

  @Override
  public void onFramePrepared(
      int frameNumber,
      CloseableReference<Bitmap> bitmapReference,
      @BitmapAnimationBackend.FrameType int frameType) {
  }

  @Override
  public void setFrameCacheListener(FrameCacheListener frameCacheListener) {
    mFrameCacheListener = frameCacheListener;
  }

  private synchronized void closeAndResetLastBitmapReference() {
    if (mFrameCacheListener != null && mLastFrameNumber != FRAME_NUMBER_UNSET) {
      mFrameCacheListener.onFrameEvicted(this, mLastFrameNumber);
    }
    CloseableReference.closeSafely(mLastBitmapReference);
    mLastBitmapReference = null;
    mLastFrameNumber = FRAME_NUMBER_UNSET;
  }
}

File: animated-drawable/src/main/java/com/facebook/fresco/animation/bitmap/cache/KeepLastFrameCache.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.animation.bitmap.cache

import android.graphics.Bitmap
import com.facebook.common.references.CloseableReference
import com.facebook.fresco.animation.bitmap.BitmapAnimationBackend.FrameType
import com.facebook.fresco.animation.bitmap.BitmapFrameCache
import com.facebook.fresco.animation.bitmap.BitmapFrameCache.FrameCacheListener
import com.facebook.imageutils.BitmapUtil
import javax.annotation.concurrent.GuardedBy

/** Simple bitmap cache that keeps the last frame and reuses it if possible. */
class KeepLastFrameCache : BitmapFrameCache {

  private var lastFrameNumber = FRAME_NUMBER_UNSET
  private var frameCacheListener: FrameCacheListener? = null

  @GuardedBy("this") private var lastBitmapReference: CloseableReference<Bitmap>? = null

  @Synchronized
  override fun getCachedFrame(frameNumber: Int): CloseableReference<Bitmap>? =
      if (lastFrameNumber == frameNumber) {
        CloseableReference.cloneOrNull(lastBitmapReference)
      } else {
        null
      }

  @Synchronized
  override fun getFallbackFrame(frameNumber: Int): CloseableReference<Bitmap>? =
      CloseableReference.cloneOrNull(lastBitmapReference)

  @Synchronized
  override fun getBitmapToReuseForFrame(
      frameNumber: Int,
      width: Int,
      height: Int
  ): CloseableReference<Bitmap>? =
      try {
        CloseableReference.cloneOrNull(lastBitmapReference)
      } finally {
        closeAndResetLastBitmapReference()
      }

  @Synchronized
  override fun contains(frameNumber: Int): Boolean =
      frameNumber == lastFrameNumber && CloseableReference.isValid(lastBitmapReference)

  @get:Synchronized
  override val sizeInBytes: Int
    get() =
        if (lastBitmapReference == null) 0
        else BitmapUtil.getSizeInBytes(lastBitmapReference!!.get())

  @Synchronized
  override fun clear() {
    closeAndResetLastBitmapReference()
  }

  @Synchronized
  override fun onFrameRendered(
      frameNumber: Int,
      bitmapReference: CloseableReference<Bitmap>,
      @FrameType frameType: Int
  ) {
    if (lastBitmapReference != null && bitmapReference.get() == lastBitmapReference?.get()) {
      return
    }
    CloseableReference.closeSafely(lastBitmapReference)
    if (lastFrameNumber != FRAME_NUMBER_UNSET) {
      frameCacheListener?.onFrameEvicted(this, lastFrameNumber)
    }
    lastBitmapReference = CloseableReference.cloneOrNull(bitmapReference)
    frameCacheListener?.onFrameCached(this, frameNumber)
    lastFrameNumber = frameNumber
  }

  override fun onFramePrepared(
      frameNumber: Int,
      bitmapReference: CloseableReference<Bitmap>,
      @FrameType frameType: Int
  ) = Unit

  override fun setFrameCacheListener(frameCacheListener: FrameCacheListener?) {
    this.frameCacheListener = frameCacheListener
  }

  @Synchronized
  private fun closeAndResetLastBitmapReference() {
    if (lastFrameNumber != FRAME_NUMBER_UNSET) {
      frameCacheListener?.onFrameEvicted(this, lastFrameNumber)
    }
    CloseableReference.closeSafely(lastBitmapReference)
    lastBitmapReference = null
    lastFrameNumber = FRAME_NUMBER_UNSET
  }

  companion object {
    private const val FRAME_NUMBER_UNSET = -1
  }
}
